File: src/renderCsrLeaderboard.ts
/**
 * Renders the TrueSkill 2 standings as a PNG in the SAME style as the ELO
 * leaderboard (renderLeaderboard.ts) so the two boards read as a set: same dark
 * carnage-screen background, centred headline, light-blue column headers, neutral
 * steel rows with a darker rank cell, and the 🥇🥈🥉 podium medals floating in
 * the left gutter. The rating column shows the rank LABEL ("diamond 5") + the
 * Halo 5 division EMBLEM + the CSR number; the rest (W-L-D / Win% / K/D) match ELO.
 *
 * Layout is computed per-render: the CSR column is sized to exactly its widest
 * row (label + emblem + number) so there's no dead space, and the three stat
 * columns evenly fill whatever width is left.
 */

import { createCanvas, loadImage, type Image, type SKRSContext2D } from "@napi-rs/canvas";
import { readFile } from "node:fs/promises";
import { displayName } from "./aliases.ts";
import { csrFromSkill, CHAMPION_THRESHOLD } from "./csr.ts";
import { csrEmblems } from "./csrEmblems.ts";
import { FONT, FONT_BOLD } from "./fonts.ts";

/** One row of a CSR board (already ranked best-first by the caller). */
export interface CsrRow {
  gamertag: string;
  skill: number; // mu - 3*sigma
  peakSkill: number; // highest skill ever held (peak CSR)
  wins: number;
  losses: number;
  draws: number;
  games: number;
  kills: number;
  deaths: number;
}

/** One CSR board table, e.g. { title: "4V4 LEADERBOARD", rows: [...] }. */
export interface CsrBoardSection {
  title: string;
  rows: CsrRow[];
}

// --- palette / layout (mirrors renderLeaderboard.ts) -------------------------

const W = 1500;
const MARGIN = 16;
const TITLE_BASELINE = 60;
const ROW_H = 46;
const ROW_GAP = 3;
const BOTTOM_PAD = 26;

const ROW_COLOR = "#39434f";
const RANK_CELL_COLOR = "#272e37";

const SECTION_TITLE_H = 56;
const HEADER_GAP = 34;
const HEADER_TO_ROWS = 12;
const EMPTY_H = 36;
const GUTTER_W = 56;
const RANK_W = 64;
const MEDAL_SIZE = 32;
const CSR_EMBLEM_H = 34; // division emblem height inside a row
const NAME_W = 340; // gamertag column width
const CSR_PAD = 20; // padding each side of the CSR content within its column
const GROUP_GAP = 10; // gap between label / emblem / number
const STAT_LABELS = ["W-L-D", "WIN%", "K/D", "PEAK CSR"] as const;

/**
 * Champion = the #1 row on a board whose CSR has cleared the Champion floor.
 * It's position-based, so it depends on the row's rank, not just its skill.
 */
function isChampion(rankIndex: number, skill: number): boolean {
  return rankIndex === 0 && csrFromSkill(skill).value >= CHAMPION_THRESHOLD;
}

/** The rank label: lowercase "tier sub" / "onyx", or "champion" for the #1≥floor. */
function rankLabel(skill: number, rankIndex = -1): string {
  if (isChampion(rankIndex, skill)) return "champion";
  const c = csrFromSkill(skill);
  return c.isOnyx ? "onyx" : `${c.tier.toLowerCase()} ${c.sub}`;
}

/** Emblem key for a row: the Halo 5 Champion insignia for #1≥floor, else the tier. */
function rowEmblem(skill: number, rankIndex: number): string {
  return isChampion(rankIndex, skill) ? "champion" : csrFromSkill(skill).emblem;
}

interface Layout {
  rankX: number; // left of the rank cell
  nameLeft: number; // left of the gamertag column
  csrLeft: number;
  csrRight: number;
  csrCenter: number;
  stats: { label: string; left: number; right: number; headerX: number; valueX: number }[];
}

/**
 * Compute column geometry: the CSR column is exactly as wide as its widest row's
 * content (label + emblem + number) plus padding; the three stat columns split
 * the remaining width to the right margin evenly.
 */
function computeLayout(
  ctx: SKRSContext2D,
  sections: CsrBoardSection[],
  limit: number,
  emblems: Record<string, Image>,
): Layout {
  ctx.font = `22px ${FONT}`;
  let maxGroup = 0;
  for (const s of sections) {
    const rows = s.rows.slice(0, limit);
    for (let i = 0; i < rows.length; i++) {
      const r = rows[i];
      const c = csrFromSkill(r.skill);
      const img = emblems[rowEmblem(r.skill, i)];
      const ew = img ? (img.width / img.height) * CSR_EMBLEM_H : 0;
      const w =
        ctx.measureText(rankLabel(r.skill, i)).width +
        GROUP_GAP + ew + GROUP_GAP +
        ctx.measureText(`(${c.value})`).width;
      maxGroup = Math.max(maxGroup, w);
    }
  }

  const rankX = MARGIN + GUTTER_W;
  const nameLeft = rankX + RANK_W;
  const csrLeft = nameLeft + NAME_W;
  const csrRight = csrLeft + maxGroup + CSR_PAD * 2;
  const statsLeft = csrRight;
  const statW = (W - MARGIN - statsLeft) / STAT_LABELS.length;
  const stats = STAT_LABELS.map((label, i) => {
    const left = statsLeft + i * statW;
    const right = left + statW;
    return { label, left, right, headerX: left + 14, valueX: right - 8 };
  });

  return { rankX, nameLeft, csrLeft, csrRight, csrCenter: (csrLeft + csrRight) / 2, stats };
}

// --- baked images ------------------------------------------------------------

let medalCache: Promise<Image[]> | undefined;
function medals(): Promise<Image[]> {
  medalCache ??= Promise.all(
    [1, 2, 3].map(async (n) =>
      loadImage(await readFile(new URL(`../assets/medal-${n}.png`, import.meta.url))),
    ),
  );
  return medalCache;
}

// --- renderer ----------------------------------------------------------------

function sectionHeight(s: CsrBoardSection, limit: number): number {
  const rows = Math.min(s.rows.length, limit);
  const body = rows ? HEADER_GAP + HEADER_TO_ROWS + rows * (ROW_H + ROW_GAP) - ROW_GAP : EMPTY_H;
  return SECTION_TITLE_H + body;
}

export async function renderCsrLeaderboardPng(
  sections: CsrBoardSection[],
  limit = 20,
): Promise<Buffer> {
  const [medalImgs, emblems] = await Promise.all([medals(), csrEmblems()]);
  const SUBTITLE_GAP = 32; // title baseline -> subtitle baseline
  const sectionsTop = TITLE_BASELINE + SUBTITLE_GAP + 14;
  const height =
    sectionsTop + sections.reduce((h, s) => h + sectionHeight(s, limit), 0) + BOTTOM_PAD;
  const canvas = createCanvas(W, height);
  const ctx = canvas.getContext("2d");
  const layout = computeLayout(ctx, sections, limit, emblems);

  // Background: near-black with a faint cool gradient like the in-game scene.
  const bg = ctx.createLinearGradient(0, 0, 0, height);
  bg.addColorStop(0, "#14171c");
  bg.addColorStop(1, "#0a0c10");
  ctx.fillStyle = bg;
  ctx.fillRect(0, 0, W, height);

  // Centred two-line headline: "CSR STANDINGS" with "HALO 3 CUSTOMS" beneath it.
  ctx.textAlign = "center";
  ctx.font = `44px ${FONT_BOLD}`;
  ctx.fillStyle = "#ffffff";
  ctx.fillText("CSR STANDINGS", W / 2, TITLE_BASELINE);
  ctx.font = `28px ${FONT}`;
  ctx.fillStyle = "#d4dbe4";
  ctx.fillText("HALO 3 CUSTOMS", W / 2, TITLE_BASELINE + SUBTITLE_GAP);
  ctx.textAlign = "left";

  let y = sectionsTop;
  for (const s of sections) {
    drawSection(ctx, s, y, limit, medalImgs, emblems, layout);
    y += sectionHeight(s, limit);
  }

  return canvas.toBuffer("image/png");
}

function drawSection(
  ctx: SKRSContext2D,
  s: CsrBoardSection,
  top: number,
  limit: number,
  medalImgs: Image[],
  emblems: Record<string, Image>,
  layout: Layout,
): void {
  const titleBaseline = top + SECTION_TITLE_H;
  ctx.font = `30px ${FONT_BOLD}`;
  ctx.fillStyle = "#ffffff";
  ctx.fillText(s.title, MARGIN, titleBaseline);

  if (!s.rows.length) {
    ctx.font = `22px ${FONT}`;
    ctx.fillStyle = "#8b95a1";
    ctx.fillText("NO MATCHES YET", MARGIN + 2, titleBaseline + EMPTY_H);
    return;
  }

  const headerBaseline = titleBaseline + HEADER_GAP;
  ctx.font = `20px ${FONT}`;
  ctx.fillStyle = "#76b5d8";
  ctx.textAlign = "left";
  ctx.fillText("#", MARGIN + GUTTER_W + 18, headerBaseline);
  ctx.fillText("PLAYERS", layout.nameLeft + 16, headerBaseline);
  ctx.fillText("CSR", layout.csrLeft + 14, headerBaseline); // left-aligned, like the others
  for (const st of layout.stats) ctx.fillText(st.label, st.headerX, headerBaseline);

  const rowsTop = headerBaseline + HEADER_TO_ROWS;
  const rows = s.rows.slice(0, limit);
  for (let i = 0; i < rows.length; i++) {
    drawRow(ctx, rows[i], i, rowsTop + i * (ROW_H + ROW_GAP), medalImgs, emblems, layout);
  }
}

function drawRow(
  ctx: SKRSContext2D,
  r: CsrRow,
  i: number,
  y: number,
  medalImgs: Image[],
  emblems: Record<string, Image>,
  layout: Layout,
): void {
  const rowX = layout.rankX;
  ctx.fillStyle = ROW_COLOR;
  ctx.fillRect(rowX, y, W - MARGIN - rowX, ROW_H);

  // Rank cell stays neutral, like the ELO board.
  ctx.fillStyle = RANK_CELL_COLOR;
  ctx.fillRect(rowX, y, RANK_W, ROW_H);

  // Dark separators at the column boundaries.
  ctx.fillStyle = "rgba(0,0,0,0.55)";
  for (const x of [rowX + RANK_W, layout.csrLeft, ...layout.stats.map((s) => s.left)]) {
    ctx.fillRect(x, y, 2, ROW_H);
  }

  const mid = y + ROW_H / 2 + 8; // baseline that visually centres 22px text

  // Podium medals float on the background gutter, outside the row.
  if (i < 3) {
    ctx.drawImage(medalImgs[i], MARGIN + (GUTTER_W - MEDAL_SIZE) / 2,
      y + (ROW_H - MEDAL_SIZE) / 2, MEDAL_SIZE, MEDAL_SIZE);
  }

  ctx.font = `22px ${FONT}`;
  ctx.fillStyle = "#ffffff";
  ctx.textAlign = "right";
  ctx.fillText(String(i + 1), rowX + RANK_W - 18, mid);
  ctx.textAlign = "left";
  ctx.fillText(displayName(r.gamertag), layout.nameLeft + 16, mid);

  // CSR cell: emblem + rank label ("diamond 5", montage colour) + CSR number
  // (white), centred as one group in the (content-sized) CSR column. Emblem leads
  // (icon-then-label) so the divisions line up cleanly down the column.
  const cell = csrFromSkill(r.skill);
  const labelText = rankLabel(r.skill, i);
  const valueText = `(${cell.value})`;
  const labelW = ctx.measureText(labelText).width;
  const valueW = ctx.measureText(valueText).width;
  const img = emblems[rowEmblem(r.skill, i)];
  const ew = img ? (img.width / img.height) * CSR_EMBLEM_H : 0;
  let gx = layout.csrCenter - (ew + GROUP_GAP + labelW + GROUP_GAP + valueW) / 2;
  if (img) ctx.drawImage(img, gx, y + (ROW_H - CSR_EMBLEM_H) / 2, ew, CSR_EMBLEM_H);
  gx += ew + GROUP_GAP;
  ctx.fillStyle = "#cfe3f2"; // montage label colour (Champion included — emblem carries the distinction)
  ctx.fillText(labelText, gx, mid);
  gx += labelW + GROUP_GAP;
  ctx.fillStyle = "#ffffff";
  ctx.fillText(valueText, gx, mid);

  // Remaining columns match the ELO board (values right-aligned in their cells).
  const winPct = r.games ? `${Math.round((r.wins / r.games) * 100)}%` : "—";
  const kd = r.deaths ? (r.kills / r.deaths).toFixed(2) : r.kills.toFixed(2);
  const peak = String(csrFromSkill(r.peakSkill).value);
  const values = [`${r.wins}-${r.losses}-${r.draws}`, winPct, kd, peak];
  ctx.textAlign = "right";
  for (let s = 0; s < layout.stats.length; s++) ctx.fillText(values[s], layout.stats[s].valueX, mid);
  ctx.textAlign = "left";
}

File: cpp/src/render_csr_leaderboard.cpp
#include "render_csr_leaderboard.h"

#include <windows.h>
#include <objidl.h>

#include <algorithm>
namespace Gdiplus {
using std::max;
using std::min;
}  // namespace Gdiplus
#include <gdiplus.h>

#include <cmath>
#include <memory>
#include <mutex>
#include <stdexcept>
#include <string>

#include "aliases.h"
#include "csr.h"
#include "csr_gfx.h"
#include "medal_assets.h"
#include "trueskill2.h"
#include "util.h"

#pragma comment(lib, "gdiplus.lib")

namespace {

using namespace Gdiplus;

// --- palette / layout (mirrors src/renderCsrLeaderboard.ts) ------------------
constexpr int W = 1500;
constexpr int MARGIN = 16;
constexpr float TITLE_BASELINE = 60.0f;
constexpr int ROW_H = 46;
constexpr int ROW_GAP = 3;
constexpr int BOTTOM_PAD = 26;

const Color ROW_COLOR(0x39, 0x43, 0x4f);
const Color RANK_CELL_COLOR(0x27, 0x2e, 0x37);

constexpr int SECTION_TITLE_H = 56;
constexpr int HEADER_GAP = 34;
constexpr int HEADER_TO_ROWS = 12;
constexpr int EMPTY_H = 36;
constexpr int GUTTER_W = 56;
constexpr int RANK_W = 64;
constexpr int MEDAL_SIZE = 32;
constexpr float CSR_EMBLEM_H = 34.0f;
constexpr int NAME_W = 340;
constexpr float CSR_PAD = 20.0f;
constexpr float GROUP_GAP = 10.0f;
constexpr float SUBTITLE_GAP = 32.0f;

const wchar_t* STAT_LABELS[4] = {L"W-L-D", L"WIN%", L"K/D", L"PEAK CSR"};

// --- small helpers -----------------------------------------------------------
std::wstring widen(const std::string& utf8) {
    if (utf8.empty()) return L"";
    int n = MultiByteToWideChar(CP_UTF8, 0, utf8.c_str(), -1, nullptr, 0);
    std::wstring w(n > 0 ? n - 1 : 0, L'\0');
    if (n > 1) MultiByteToWideChar(CP_UTF8, 0, utf8.c_str(), -1, w.data(), n);
    return w;
}

void ensureGdiplus() {
    static std::once_flag once;
    static Status startStatus = Ok;
    std::call_once(once, [] {
        GdiplusStartupInput in;
        ULONG_PTR token = 0;  // intentionally never shut down (process lifetime)
        startStatus = GdiplusStartup(&token, &in, nullptr);
    });
    if (startStatus != Ok) throw std::runtime_error("GdiplusStartup failed");
}

CLSID pngEncoderClsid() {
    UINT count = 0, bytes = 0;
    if (GetImageEncodersSize(&count, &bytes) != Ok || !bytes)
        throw std::runtime_error("no GDI+ image encoders");
    std::vector<unsigned char> buf(bytes);
    auto* infos = reinterpret_cast<ImageCodecInfo*>(buf.data());
    if (GetImageEncoders(count, bytes, infos) != Ok)
        throw std::runtime_error("GetImageEncoders failed");
    for (UINT i = 0; i < count; ++i)
        if (wcscmp(infos[i].MimeType, L"image/png") == 0) return infos[i].Clsid;
    throw std::runtime_error("PNG encoder not found");
}

struct PxFont {
    std::unique_ptr<Font> font;
    float ascent;
    PxFont(BlenderFace face, float px) {
        const FontFamily* ff = face.family;
        font = std::make_unique<Font>(ff, px, face.style, UnitPixel);
        ascent = px * ff->GetCellAscent(face.style) / ff->GetEmHeight(face.style);
    }
};

void drawText(Graphics& g, const std::wstring& s, const PxFont& f, const Brush& brush, float x,
              float baseline, const StringFormat* fmt) {
    g.DrawString(s.c_str(), -1, f.font.get(), PointF(x, baseline - f.ascent), fmt, &brush);
}

float measure(Graphics& g, const std::wstring& s, const PxFont& f, const StringFormat* fmt) {
    RectF box;
    g.MeasureString(s.c_str(), -1, f.font.get(), PointF(0, 0), fmt, &box);
    return box.Width;
}

void drawTextRight(Graphics& g, const std::wstring& s, const PxFont& f, const Brush& brush,
                   float right, float baseline, const StringFormat* fmt) {
    drawText(g, s, f, brush, right - measure(g, s, f, fmt), baseline, fmt);
}

void drawTextCenter(Graphics& g, const std::wstring& s, const PxFont& f, const Brush& brush,
                    float center, float baseline, const StringFormat* fmt) {
    drawText(g, s, f, brush, center - measure(g, s, f, fmt) / 2.0f, baseline, fmt);
}

std::unique_ptr<Bitmap> loadMedal(const MedalAsset& asset) {
    HGLOBAL hg = GlobalAlloc(GMEM_MOVEABLE, asset.size);
    if (!hg) throw std::runtime_error("GlobalAlloc failed");
    memcpy(GlobalLock(hg), asset.data, asset.size);
    GlobalUnlock(hg);
    IStream* stream = nullptr;
    if (FAILED(CreateStreamOnHGlobal(hg, TRUE, &stream))) {
        GlobalFree(hg);
        throw std::runtime_error("CreateStreamOnHGlobal failed");
    }
    auto bmp = std::make_unique<Bitmap>(stream);
    stream->Release();
    if (bmp->GetLastStatus() != Ok) throw std::runtime_error("medal PNG decode failed");
    return bmp;
}

// --- CSR row helpers (mirror renderCsrLeaderboard.ts) ------------------------
bool isChampion(int rankIndex, double skill) {
    return rankIndex == 0 && csrFromSkill(skill).value >= CHAMPION_THRESHOLD;
}
std::string rankLabel(double skill, int rankIndex) {
    if (isChampion(rankIndex, skill)) return "champion";
    Csr c = csrFromSkill(skill);
    return c.isOnyx ? "onyx" : util::toLower(c.tier) + " " + std::to_string(c.sub);
}
std::string rowEmblemKey(double skill, int rankIndex) {
    return isChampion(rankIndex, skill) ? "champion" : csrFromSkill(skill).emblem;
}

float emblemWidth(Bitmap* img) {
    if (!img || img->GetHeight() == 0) return 0;
    return static_cast<float>(img->GetWidth()) / static_cast<float>(img->GetHeight()) * CSR_EMBLEM_H;
}

struct StatCol {
    float left, right, headerX, valueX;
};
struct Layout {
    float rankX, nameLeft, csrLeft, csrRight, csrCenter;
    StatCol stats[4];
};

Layout computeLayout(Graphics& g, const std::vector<CsrBoardSection>& sections, size_t limit,
                     const PxFont& rowFont, const StringFormat* fmt) {
    float maxGroup = 0;
    for (const auto& s : sections) {
        size_t n = std::min(s.rows.size(), limit);
        for (size_t i = 0; i < n; ++i) {
            const CsrRow& r = s.rows[i];
            Csr c = csrFromSkill(r.skill);
            Bitmap* img = csrEmblem(rowEmblemKey(r.skill, static_cast<int>(i)));
            float ew = emblemWidth(img);
            float w = measure(g, widen(rankLabel(r.skill, static_cast<int>(i))), rowFont, fmt) +
                      GROUP_GAP + ew + GROUP_GAP +
                      measure(g, widen("(" + std::to_string(c.value) + ")"), rowFont, fmt);
            maxGroup = std::max(maxGroup, w);
        }
    }

    Layout L;
    L.rankX = MARGIN + GUTTER_W;
    L.nameLeft = L.rankX + RANK_W;
    L.csrLeft = L.nameLeft + NAME_W;
    L.csrRight = L.csrLeft + maxGroup + CSR_PAD * 2;
    L.csrCenter = (L.csrLeft + L.csrRight) / 2.0f;
    float statsLeft = L.csrRight;
    float statW = (W - MARGIN - statsLeft) / 4.0f;
    for (int i = 0; i < 4; ++i) {
        float left = statsLeft + i * statW;
        float right = left + statW;
        L.stats[i] = {left, right, left + 14, right - 8};
    }
    return L;
}

int sectionHeight(const CsrBoardSection& s, size_t limit) {
    size_t rows = std::min(s.rows.size(), limit);
    int body = rows ? HEADER_GAP + HEADER_TO_ROWS + static_cast<int>(rows) * (ROW_H + ROW_GAP) -
                          ROW_GAP
                    : EMPTY_H;
    return SECTION_TITLE_H + body;
}

}  // namespace

CsrBoardSection buildCsrBoardSection(const std::vector<StoredMatch>& matches, Category cat) {
    std::vector<StoredMatch> ms;
    for (const auto& m : matches)
        if (boardCategory(m) == cat) ms.push_back(m);

    std::vector<MMR> ratings = rateCategory(ms);
    std::vector<const MMR*> ranked;
    for (const auto& r : ratings)
        if (r.games > 0) ranked.push_back(&r);
    std::stable_sort(ranked.begin(), ranked.end(),
                     [](const MMR* a, const MMR* b) { return a->skill > b->skill; });

    CsrBoardSection section;
    section.title = std::string(categoryLabel(cat)) + " LEADERBOARD";
    // categoryLabel already upper-cases FFA; 2v2/4v4 stay as-is, matching the TS
    // title CATEGORY_LABEL[c].toUpperCase() (2v2/4v4 have no letters to upper).
    for (const MMR* r : ranked) {
        CsrRow row;
        row.gamertag = r->gamertag;
        row.skill = r->skill;
        row.peakSkill = r->peakSkill;
        row.wins = r->wins;
        row.losses = r->losses;
        row.draws = r->draws;
        row.games = r->games;
        row.kills = r->kills;
        row.deaths = r->deaths;
        section.rows.push_back(std::move(row));
    }
    return section;
}

std::vector<CsrBoardSection> buildCsrBoardSections(const std::vector<StoredMatch>& matches) {
    std::vector<CsrBoardSection> out;
    for (Category c : BOARD_CATEGORIES) out.push_back(buildCsrBoardSection(matches, c));
    return out;
}

std::vector<std::uint8_t> renderCsrLeaderboardPng(const std::vector<CsrBoardSection>& sections,
                                                  size_t limit) {
    ensureGdiplus();

    float sectionsTop = TITLE_BASELINE + SUBTITLE_GAP + 14;
    int height = static_cast<int>(sectionsTop) + BOTTOM_PAD;
    for (const auto& s : sections) height += sectionHeight(s, limit);

    Bitmap bmp(W, height, PixelFormat32bppARGB);
    Graphics g(&bmp);
    g.SetTextRenderingHint(TextRenderingHintAntiAliasGridFit);
    g.SetSmoothingMode(SmoothingModeHighQuality);
    g.SetInterpolationMode(InterpolationModeHighQualityBicubic);

    const StringFormat* fmt = StringFormat::GenericTypographic();

    {
        LinearGradientBrush bg(PointF(0, 0), PointF(0, static_cast<float>(height)),
                               Color(0x14, 0x17, 0x1c), Color(0x0a, 0x0c, 0x10));
        g.FillRectangle(&bg, 0, 0, W, height);
    }

    PxFont titleFont(blenderFace(true), 44);
    PxFont subtitleFont(blenderFace(false), 28);
    PxFont sectionFont(blenderFace(true), 30);
    PxFont headerFont(blenderFace(false), 20);
    PxFont rowFont(blenderFace(false), 22);

    SolidBrush white(Color(0xff, 0xff, 0xff));
    SolidBrush subtitleBrush(Color(0xd4, 0xdb, 0xe4));
    SolidBrush headerBrush(Color(0x76, 0xb5, 0xd8));
    SolidBrush emptyBrush(Color(0x8b, 0x95, 0xa1));
    SolidBrush rowBrush(ROW_COLOR);
    SolidBrush rankBrush(RANK_CELL_COLOR);
    SolidBrush divider(Color(140, 0, 0, 0));
    SolidBrush csrLabelBrush(Color(0xcf, 0xe3, 0xf2));

    std::unique_ptr<Bitmap> medals[3] = {loadMedal(MEDAL_ASSETS[0]), loadMedal(MEDAL_ASSETS[1]),
                                         loadMedal(MEDAL_ASSETS[2])};

    Layout L = computeLayout(g, sections, limit, rowFont, fmt);

    // Centred two-line headline.
    drawTextCenter(g, L"CSR STANDINGS", titleFont, white, W / 2.0f, TITLE_BASELINE, fmt);
    drawTextCenter(g, L"HALO 3 CUSTOMS", subtitleFont, subtitleBrush, W / 2.0f,
                   TITLE_BASELINE + SUBTITLE_GAP, fmt);

    float top = sectionsTop;
    for (const auto& s : sections) {
        float titleBaseline = top + SECTION_TITLE_H;
        drawText(g, widen(s.title), sectionFont, white, MARGIN, titleBaseline, fmt);

        if (s.rows.empty()) {
            drawText(g, L"NO MATCHES YET", rowFont, emptyBrush, MARGIN + 2, titleBaseline + EMPTY_H,
                     fmt);
            top += sectionHeight(s, limit);
            continue;
        }

        float headerBaseline = titleBaseline + HEADER_GAP;
        drawText(g, L"#", headerFont, headerBrush, MARGIN + GUTTER_W + 18, headerBaseline, fmt);
        drawText(g, L"PLAYERS", headerFont, headerBrush, L.nameLeft + 16, headerBaseline, fmt);
        drawText(g, L"CSR", headerFont, headerBrush, L.csrLeft + 14, headerBaseline, fmt);
        for (int c = 0; c < 4; ++c)
            drawText(g, STAT_LABELS[c], headerFont, headerBrush, L.stats[c].headerX, headerBaseline,
                     fmt);

        float rowsTop = headerBaseline + HEADER_TO_ROWS;
        size_t n = std::min(s.rows.size(), limit);
        for (size_t i = 0; i < n; ++i) {
            const CsrRow& r = s.rows[i];
            float y = rowsTop + static_cast<float>(i) * (ROW_H + ROW_GAP);
            float rowX = L.rankX;

            g.FillRectangle(&rowBrush, rowX, y, static_cast<float>(W - MARGIN) - rowX,
                            static_cast<float>(ROW_H));
            g.FillRectangle(&rankBrush, rowX, y, static_cast<float>(RANK_W),
                            static_cast<float>(ROW_H));

            // Dark separators: rank cell, CSR column, each stat column.
            g.FillRectangle(&divider, rowX + RANK_W, y, 2.0f, static_cast<float>(ROW_H));
            g.FillRectangle(&divider, L.csrLeft, y, 2.0f, static_cast<float>(ROW_H));
            for (int c = 0; c < 4; ++c)
                g.FillRectangle(&divider, L.stats[c].left, y, 2.0f, static_cast<float>(ROW_H));

            float mid = y + ROW_H / 2.0f + 8.0f;

            if (i < 3)
                g.DrawImage(medals[i].get(),
                            RectF(MARGIN + (GUTTER_W - MEDAL_SIZE) / 2.0f,
                                  y + (ROW_H - MEDAL_SIZE) / 2.0f, MEDAL_SIZE, MEDAL_SIZE));

            drawTextRight(g, std::to_wstring(i + 1), rowFont, white, rowX + RANK_W - 18, mid, fmt);
            drawText(g, widen(displayName(r.gamertag)), rowFont, white, L.nameLeft + 16, mid, fmt);

            // CSR cell: emblem + rank label + (value), centred as one group.
            Csr cell = csrFromSkill(r.skill);
            std::wstring labelText = widen(rankLabel(r.skill, static_cast<int>(i)));
            std::wstring valueText = widen("(" + std::to_string(cell.value) + ")");
            float labelW = measure(g, labelText, rowFont, fmt);
            float valueW = measure(g, valueText, rowFont, fmt);
            Bitmap* img = csrEmblem(rowEmblemKey(r.skill, static_cast<int>(i)));
            float ew = emblemWidth(img);
            float gx = L.csrCenter - (ew + GROUP_GAP + labelW + GROUP_GAP + valueW) / 2.0f;
            if (img) {
                g.DrawImage(img, RectF(gx, y + (ROW_H - CSR_EMBLEM_H) / 2.0f, ew, CSR_EMBLEM_H));
                gx += ew + GROUP_GAP;
            }
            drawText(g, labelText, rowFont, csrLabelBrush, gx, mid, fmt);
            gx += labelW + GROUP_GAP;
            drawText(g, valueText, rowFont, white, gx, mid, fmt);

            // Stat columns (right-aligned values).
            std::string winPct =
                r.games ? std::to_string(util::jsRound(static_cast<double>(r.wins) /
                                                       static_cast<double>(r.games) * 100.0)) +
                              "%"
                        : "\xE2\x80\x94";
            std::string kd = r.deaths ? util::toFixed2(static_cast<double>(r.kills) /
                                                       static_cast<double>(r.deaths))
                                      : util::toFixed2(static_cast<double>(r.kills));
            std::string peak = std::to_string(csrFromSkill(r.peakSkill).value);
            const std::wstring values[4] = {
                widen(std::to_string(r.wins) + "-" + std::to_string(r.losses) + "-" +
                      std::to_string(r.draws)),
                widen(winPct),
                widen(kd),
                widen(peak),
            };
            for (int c = 0; c < 4; ++c)
                drawTextRight(g, values[c], rowFont, white, L.stats[c].valueX, mid, fmt);
        }

        top += sectionHeight(s, limit);
    }

    IStream* stream = nullptr;
    if (FAILED(CreateStreamOnHGlobal(nullptr, TRUE, &stream)))
        throw std::runtime_error("CreateStreamOnHGlobal failed");
    CLSID png = pngEncoderClsid();
    Status saved = bmp.Save(stream, &png, nullptr);
    if (saved != Ok) {
        stream->Release();
        throw std::runtime_error("PNG encode failed");
    }

    HGLOBAL hg = nullptr;
    GetHGlobalFromStream(stream, &hg);
    SIZE_T size = GlobalSize(hg);
    void* data = GlobalLock(hg);
    std::vector<std::uint8_t> out(static_cast<const std::uint8_t*>(data),
                                  static_cast<const std::uint8_t*>(data) + size);
    GlobalUnlock(hg);
    stream->Release();
    return out;
}
